### PR TITLE
Read GISAID data files efficiently and remove duplicate entries during GISAID filtering

### LINKED ISSUE #36: Remove duplicate entries during GISAID filtering
Although it is possible to link each duplicated entry to its corresponding sequence, there's only 1049 duplicates in the 2021-07-02 GISAID release out of 2.1m entries (0.048%) and none of them are from Canada:

```
Netherlands       271
USA               186
Israel            182
France            102
Lithuania          99
Spain              45
Italy              25
South Korea        24
Japan              21
Bonaire            21
United Kingdom     19
Burkina Faso       12
Curacao             9
Belgium             9
Cuba                4
Switzerland         4
Senegal             3
Rwanda              2
Iran                1
South Africa        1
Ireland             1
Poland              1
China               1
Czech Republic      1
Turkey              1
Chile               1
```

Telling duplicates apart would require checking sequence length, GC content and N content, so bit of extra computation for a very small proportion of the data. Unfortunately, the unique GISAID ID is not present in the FASTA file sequence headers, but some date values are present by default.

So I propose that during filtering of GISAID data, all duplicate entries should be dropped so that at subsequent stages it can be assumed that there are no duplicates.

### LINKED ISSUE #37: Python tarfile getnames() method very slow for large files
Running `tarfile.TarFile.getnames()` is very slow for large files like the sequences and metadata `tar.xz` files from GISAID. Using the `tarfile.TarFile.next()` method is much faster.

image

1m26s vs 3ms for `sequences_fasta_2021_07_02.tar.xz`

5.06s vs 3ms for `metadata_tsv_2021_07_02.tar.xz`

---

```python
def get_metadata_file_from_tar(tar: tarfile.TarFile) -> Optional[IO[bytes]]:
    while True:
        tinfo = tar.next()
        if tinfo is None:
            break
        if tinfo.name.endswith('.tsv'):
            return tar.extractfile(tinfo)
```
instead of https://github.com/nhhaidee/scovtree/blob/7c331a11663acbada7cfdc99938c7915bad1ecf6/bin/filter_gisaid.py#L220-L222

---

```python
def get_fasta_from_tar(tar: tarfile.TarFile) -> Optional[IO[bytes]]:
    while True:
        tinfo = tar.next()
        if tinfo is None:
            break
        if tinfo.name.endswith('.fasta'):
            return tar.extractfile(tinfo)
```

instead of https://github.com/nhhaidee/scovtree/blob/7c331a11663acbada7cfdc99938c7915bad1ecf6/bin/filter_gisaid.py#L233

### PR DESCRIPTION
**Pull request by nhhaidee**
This PR addresses https://github.com/nhhaidee/scovtree/issues/36 and https://github.com/nhhaidee/scovtree/issues/37

## PR checklist

- [x] This comment contains a description of changes (with reason).
- [x] If you've fixed a bug or added code that should be tested, add tests!
  - [ ] If you've added a new tool - add to the software_versions process and a regex to `scrape_software_versions.py`
  - [ ] If you've added a new tool - have you followed the pipeline conventions in the [contribution docs](https://github.com/nhhaidee/scovtree/tree/master/.github/CONTRIBUTING.md)
  - [ ] If necessary, also make a PR on the nhhaidee/scovtree _branch_ on the [nhhaidee/nf-test-datasets](https://github.com/nhhaidee/nf-test-datasets) repository.
- [ ] Ensure the test suite passes (`nextflow run . -profile test,docker`).
- [ ] Usage Documentation in `docs/usage.md` is updated.
- [ ] Output Documentation in `docs/output.md` is updated.
- [x] `CHANGELOG.md` is updated.
- [ ] `README.md` is updated (including new tool citations and authors/contributors).

### CODE REVIEW

**Comment by nhhaidee** [comment]
@peterk87 `df_subset` contains duplicate entries so I added code as below to remove duplicate entries before filtering/sampling

https://github.com/nhhaidee/scovtree/blob/f5670e7472d3e03e28323e18b3dbcb918b565dd1/bin/filter_gisaid.py#L64-L67

File: bin/filter_gisaid.py [hunk 1]
```diff
@@ -265,5 +263,14 @@ def read_fasta_tarxz(tarxz_path) -> Iterator[Tuple[str, str]]:
         yield title, "".join(lines).replace(" ", "").replace("\r", "")
 
 
+def get_file_from_tar(tar: tarfile.TarFile) -> Optional[IO[bytes]]:
```
Comment (peterk87) [inline bin/filter_gisaid.py]: I like this generic function. One suggestion would be to have the name/extension as an argument to the function:

```python
def get_file_from_tar(tar: tarfile.TarFile, name: str) -> ...:
    while True:
        ...
        if re.match(name, file_member.name):
```

Function call might look like:

```python
handle = get_file_from_tar(tar, r'.*\.fasta$')
```

**Review by peterk87 — COMMENTED** [review]
Thanks @nhhaidee! The changes look good to me.

I only have one minor comment.

**Comment by peterk87** [comment]
@nhhaidee be sure to do a version bump and update the `CHANGELOG.md`

File: bin/filter_gisaid.py (hunk 1 above)
Comment (nhhaidee) [reply bin/filter_gisaid.py]: Updated by https://github.com/nhhaidee/scovtree/pull/38/commits/25fbbb5927d35e0972f4070bcc2f855d042070b9

**Comment by nhhaidee** [comment]
@peterk87 The PR contains enhancements so I will release v1.2.0.

File: CHANGELOG.md [hunk 2]
```diff
@@ -7,16 +7,15 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
...
+## [v1.1.0](https://github.com/nhhaidee/scovtree/releases/tag/1.1.0) - [July 06 2021]
 
-### `Dependencies`
+### :warning: Major enhancements
 
-### `Deprecated`
+* This release addresses an enhancement in which low abundance lineages should be better represented in all possible stages. The sampling is done separately for each unique lineage to ensure that each lineage is properly represented and more balanced in Phyolgenetic Analysis.
 
-## [v1.1.0](https://github.com/nhhaidee/scovtree/releases/tag/1.1.0) - [July 06 2021]
+## [v1.2.0](https://github.com/nhhaidee/scovtree/releases/tag/1.2.0) - [July 07 2021]
```
Comment (peterk87) [inline CHANGELOG.md]: Most recent entry should be first.

Comment (nhhaidee) [reply CHANGELOG.md]: Updated by https://github.com/nhhaidee/scovtree/pull/38/commits/db7c7401b94acb55f9290c502dd5c29abd2cfbe0

**Pull request merged**